packages/opal-client/opal_client/data/updater.py

```diff
@@ -330,216 +404,298 @@ async def stop(self):
-        """Calculate an hash (sah256) on the given data, if data isn't a
-        string, it will be converted to JSON.
+    def calc_hash(data: JsonableValue) -> str:
+        """Calculates a SHA-256 hash of the given data. If 'data' is not a
+        string, it is first serialized to JSON. Returns an empty string on
+        failure.
+
+        Args:
+            data (JsonableValue): The data to be hashed.
 
-        String are encoded as 'utf-8' prior to hash calculation.
         Returns:
-            the hash of the given data (as a a hexdigit string) or '' on failure to process.
+            str: The hexadecimal representation of the SHA-256 hash.
         """
         try:
             if not isinstance(data, str):
                 data = json.dumps(data, default=pydantic_encoder)
             return hashlib.sha256(data.encode("utf-8")).hexdigest()
-        except:
-            logger.exception("Failed to calculate hash for data {data}", data=data)
+        except Exception as e:
+            logger.exception(f"Failed to calculate hash for data {data}: {e}")
             return ""
 
-    async def _update_policy_data(
-        self,
-        update: DataUpdate,
-        store_queue_number: TakeANumberQueue.Number,
-    ):
-        """Fetches policy data (policy configuration) from backend and updates
-        it into policy-store (i.e. OPA)"""
-
-        if update is None:
-            return
-
-        # types / defaults
-        urls: List[Tuple[str, FetcherConfig, Optional[JsonableValue]]] = None
-        entries: List[DataSourceEntry] = []
-        # if we have an actual specification for the update
-        if update is not None:
-            # Check each entry's topics to only process entries designated to us
-            entries = [
-                entry
-                for entry in update.entries
-                if entry.topics
-                and not set(entry.topics).isdisjoint(set(self._data_topics))
-            ]
-            urls = []
-            for entry in entries:
-                config = entry.config
-                if self._shard_id is not None:
-                    headers = config.get("headers", {})
-                    headers.update({"X-Shard-ID": self._shard_id})
-                    config["headers"] = headers
-                urls.append((entry.url, config, entry.data))
-
-        if len(entries) > 0:
-            logger.info("Fetching policy data", urls=repr(urls))
-        else:
-            logger.warning(
-                "None of the update's entries are designated to subscribed topics"
+    async def _update_policy_data(self, update: DataUpdate) -> None:
+        """Performs the core data update process for the given DataUpdate
+        object.
+
+        Steps:
+          1. Iterate over the DataUpdate entries.
+          2. For each entry, check if any of its topics match our client's topics.
+          3. Acquire a lock for the destination path, so we don't overwrite concurrently.
```

Comment:
```suggestion
          3. Acquire a lock for the destination path, so we don't fetch and overwrite concurrently.
             - Note: This means that fetches that can technically happen concurrently wait on one another.
                          This can be improved with  a Fetcher-Writer Lock ( a la Reader-Writer Lock ) pattern
```